flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/enumerator/KafkaSourceEnumerator.java

```diff
@@ -64,6 +64,8 @@
     private final Properties properties;
     private final long partitionDiscoveryIntervalMs;
     private final SplitEnumeratorContext<KafkaPartitionSplit> context;
+    private volatile boolean partitionDiscoveryFinished = false;
```

Comment:
It seems that we can just reuse the `noMoreNewPartitionSplits` flag.

- If the periodic partition discovery is disabled, then after the first partition discovery is done, set the `noMoreNewPartitionSplits` flag to true. The subsequent `assignPendingPartitionSplits` will just send the `NoMoreSplitsEvent` to all the readers. As long as the flag is set in the main thread, the readers who registered before the first partition discovery is done will not receive duplicate `NoMoreSplitsEvent` in this case.
- Otherwise, the `noMoreNewPartitionSplits` will always be set to false, and no `NoMoreSplitsEvent` will be sent.